File: PrimeFactors.jl
macro dountil(condition, expression)
  :(while true
      $expression
      if $condition; break; end
    end) |> esc
end

mutable struct SmallCoprime
  number::Int
  coprime::Int
end

SmallCoprime(n) = SmallCoprime(n, 1)

function next!(sc::SmallCoprime)
  if sc.coprime >= sc.number - 1; return nothing; end
  @dountil gcd(sc.number, sc.coprime) == 1  sc.coprime += 1
  sc.coprime
end

function solve(number::Int)
  (p, q) = (1, number)
  coprime = SmallCoprime(number)
  @dountil p < number && q < number begin
    next!(coprime)
    period = evenperiodof(coprime)
    (p, q) = getfactors(coprime, period)
  end
  (p, q)
end

function evenperiodof(sc::SmallCoprime)
  period = 2
  @dountil (sc.coprime ^ period) % sc.number == 1  period += 2
  period
end

function getfactors(sc::SmallCoprime, period)
  α = sc.coprime ^ (period ÷ 2)
  p = gcd(sc.number, α + 1)
  q = gcd(sc.number, α - 1)
  (p, q)
end


# TEST ========================================================================
map([15, 35, 65, 91]) do n
  println("The factors of $(n) are $(solve(n))")
end

File: prime-factors.rs
macro_rules! repeat {
  ($body:block until $condition:expr) => (
    loop {
      $body
      if $condition { break }
    }
  );
}

fn gcd(a: u32, b: u32) -> u32 {
  if b == 0 { return a; }
  gcd(b, a % b)
}

struct SmallCoprime {
  number: u32,
  coprime: u32
}

impl SmallCoprime {
  fn init(number: u32) -> SmallCoprime {
    SmallCoprime{ number, coprime: 1 }
  }

  fn even_period(&self) -> u32 {
    let mut period = 2;
    repeat!({
       period += 2
    } until self.coprime.pow(period) % self.number == 1);
    period
  }

  fn factors(&self, period: u32) -> (u32, u32) {
    let w = self.coprime.pow(period / 2);
    let p = gcd(self.number, w + 1);
    let q = gcd(self.number, w - 1);
    (p, q)
  }
}

impl Iterator for SmallCoprime {
  type Item = u32;

  fn next(&mut self) -> Option<Self::Item> {
    if self.coprime >= self.number-1 { return None }
    repeat!({
      self.coprime += 1
    } until gcd(self.number, self.coprime) == 1);
    Some(self.coprime)
  }
}

fn solve(number: u32) -> (u32, u32) {
  let mut coprime = SmallCoprime::init(number);
  let mut factors: (u32, u32);
  repeat!({
    coprime.next();
    let period = coprime.even_period();
    factors = coprime.factors(period);
  } until factors.0 < number && factors.1 < number);
  factors
}

fn main() {
  for number in vec![15, 35, 65, 91] {
    println!("The factors of {} are {:?}", number, solve(number));
  }
}